terraform/aws/bucket-access.tf

```diff
@@ -0,0 +1,145 @@
+/*
+  Creates one aws_s3_bucket_policy per bucket - there can't be more than one as
+  they otherwise replace each other when applied.
+
+  The bucket policies grant bucket specific permissions to specific IAM Roles
+  based on them having `bucket_admin_access` or `bucket_readonly_access`
+  referencing the bucket via `var.hub_cloud_permissions`.
+*/
+
+locals {
+  /*
+    The bucket_role_actions local variable defined below is a list of objects
+    generated from `var.hub_cloud_permissions` roles and their respective
+    bucket_admin_access and bucket_readonly_access lists.
+
+    If for example `var.hub_cloud_permissions` is:
+
+    hub_cloud_permissions:
+      staging:
+        user-sa:
+          bucket_admin_access: [scratch-staging]
+      sciencecore:
+        user-sa:
+          bucket_admin_access: [scratch-sciencecore]
+          bucket_readonly_access: [persistent-sciencecore]
+        admin-sa:
+          bucket_admin_access: [scratch-sciencecore, persistent-sciencecore]
+
+    Then, the `local.bucket_role_actions` will look like below, with one list
+    item for each element in all `bucket_admin/readonly_access` lists:
+
+    bucket_role_actions:
+      - bucket: scratch-staging
+        role: staging
+        actions: ["s3:*"]
+      - bucket: scratch-sciencecore
+        role: sciencecore
+        actions: ["s3:*"]
+      - bucket: scratch-sciencecore
+        role: sciencecore-admin-sa
+        actions: ["s3:*"]
+      - bucket: persistent-sciencecore
+        role: sciencecore
+        actions: ["s3:ListBucket", "s3:GetObject", "s3:GetObjectVersion"]
+      - bucket: persistent-sciencecore
+        role: sciencecore
+        actions: ["s3:*"]
+  */
+  bucket_role_actions = flatten([
+    for hub, hub_value in var.hub_cloud_permissions : [
+      for role, role_value in hub_value : flatten([
```

Comment:
Can we keep the credit for the `flatten` trick from https://github.com/2i2c-org/infrastructure/pull/3932/files#diff-c1bb77fe872eda5a7e42a026322e595cf40d3595a0605c60be0793a362310639L40